### PR TITLE
add install script for mpich-3.3.1

### PR DESCRIPTION
**Pull request by raffenet**

### CODE REVIEW

**Comment by raffenet** [comment]
@zhenz I was able to test this script on a CentOS 7 image. It will build and install MPICH 3.3.1 in `/usr/local`. Should be suitable for running MPI applications on a single node. Next steps should be getting configuration right for multinode, infiniband, etc.

**Comment by zhenz** [comment]
> @zhenz I was able to test this script on a CentOS 7 image. It will build and install MPICH 3.3.1 in `/usr/local`. Should be suitable for running MPI applications on a single node. Next steps should be getting configuration right for multinode, infiniband, etc.

Thanks, @raffenet! Looks good to me! Could you please create an item for your image in Chameleon Appliance Catalog https://www.chameleoncloud.org/appliances/? I'd like you to create it so that you can edit it later; otherwise, if the item is created by me, you won't be able to edit.

**Comment by raffenet** [comment]
What do I put in the form for appliance ID for the various sites? It won't let me submit without it.

**Comment by zhenz** [comment]
> What do I put in the form for appliance ID for the various sites? It won't let me submit without it.

hmm...I thought the IDs are not required. Did you have an image created already? If so, you can use this link https://chi.uc.chameleoncloud.org/project/images to find the id. Just find the image from the list, and click the image name to get details, including id. Otherwise, try using a fake string or use the id of "mpi-graph500-generic", and we can edit it later.

**Comment by raffenet** [comment]
OK, I put in a fake one and the form was happy.

**Comment by zhenz** [comment]
> OK, I put in a fake one and the form was happy.

Thank you! I see your item in appliance catalog. I'll try to build the image using your new script and put the id in.

File: elements/mpich-3.3.1/post-install.d/05-mpich [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#############################################################
+# Pre-Installation
+#############################################################
+
+MPICH_TMP_FOLDER=/root/tmp
+mkdir -p $MPICH_TMP_FOLDER
+
+#############################################################
+# Installation (runfile)
```
Comment (zhenz) [inline elements/mpich-3.3.1/post-install.d/05-mpich]: remove "(runfile)"

File: elements/mpich-3.3.1/post-install.d/05-mpich [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#############################################################
+
+cd $MPICH_TMP_FOLDER
+
+# Get the MPICH 3.3.1
+wget -nv https://www.mpich.org/static/downloads/3.3.1/mpich-3.3.1.tar.gz
+tar xvfz mpich-3.3.1.tar.gz && cd mpich-3.3.1
+
+./configure
+
```
Comment (zhenz) [inline elements/mpich-3.3.1/post-install.d/05-mpich:31]: Need to install after configuration? "make" and "make install".

File: elements/mpich-3.3.1/post-install.d/05-mpich [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#############################################################
+# Cleaning
+#############################################################
+
+# Clean the tmp folder used by CUDA
```
Comment (zhenz) [inline elements/mpich-3.3.1/post-install.d/05-mpich]: Remove "used by CUDA"

File: elements/mpich-3.3.1/post-install.d/05-mpich [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+rm -rf $MPICH_TMP_FOLDER
```
Comment (zhenz) [inline elements/mpich-3.3.1/post-install.d/05-mpich:41]: Can we keep some examples, like hello_world, somewhere before deleting the installation directory?

**Review by zhenz — CHANGES_REQUESTED** [review]
Hi, @raffenet! The script doesn't actually install MPICH after configuration. Also some minor changes about the comments in the script. Thanks!

File: elements/mpich-3.3.1/post-install.d/05-mpich (hunk 4 above)
Comment (raffenet) [reply elements/mpich-3.3.1/post-install.d/05-mpich:41]: I added a step to copy the `cpi` example that is built with MPICH in `/usr/local/bin`. It's also possible to keep the whole `examples` directory from within the MPICH source package if we want.

Comment (zhenz) [reply elements/mpich-3.3.1/post-install.d/05-mpich:41]: One example should be good enough. We'll see if users ask more in the future. Thank you for the quick fixes, and I'll try deploy again!

**Review by zhenz — APPROVED** [review]

**Comment by zhenz** [comment]
@raffenet I updated the installation folder from "/root/tmp" to "/root" and didn't clean the folder at the end. Because the cpi program required libs and compiled scripts are stored in there. We need those to run the cpi example. Let me know if it's okay.

I tested and deployed the image at both Chameleon sites. And made some changes here https://www.chameleoncloud.org/appliances/71/. 

Could you take a look at the script and https://www.chameleoncloud.org/appliances/71/? If everything looks good to you, I'll merge the branch and publish the item in Appliance Catalog. 

Thank you!

**Comment by raffenet** [comment]
> @raffenet I updated the installation folder from "/root/tmp" to "/root" and didn't clean the folder at the end. Because the cpi program required libs and compiled scripts are stored in there. We need those to run the cpi example. Let me know if it's okay.

That makes sense. Libtool started linking the example differently in CentOS 7, but it should be fine to keep these around.

> 
> I tested and deployed the image at both Chameleon sites. And made some changes here https://www.chameleoncloud.org/appliances/71/.
> 
> Could you take a look at the script and https://www.chameleoncloud.org/appliances/71/? If everything looks good to you, I'll merge the branch and publish the item in Appliance Catalog.

I wasn't able to launch an instance earlier because there were no nodes available. I will keep trying and get back to you.

> 
> Thank you!

**Comment by zhenz** [comment]
If you see the error saying no nodes available when creating instances, that's a launch failure and might be caused by a broken node. You can create another lease with a different node and retry. But let me know if you experience any issue using Chameleon.

**Comment by raffenet** [comment]
Oh, OK. I was able to launch successfully now. However there is still a problem with the example. I'm going to modify the script to build the example separately, as that works in my testing.

**Comment by raffenet** [comment]
@zhenz just pushed new version that should install a working `cpi` example.

**Comment by raffenet** [comment]
You can confirm this by running the following in the appliance:

> mpiexec -n 4 cpi

Output should look like:

> [cc@mpich ~]$ mpiexec -n 4 cpi
> Process 0 of 4 is on mpich.novalocal
> Process 1 of 4 is on mpich.novalocal
> Process 3 of 4 is on mpich.novalocal
> Process 2 of 4 is on mpich.novalocal
> pi is approximately 3.1415926544231239, Error is 0.0000000008333307
> wall clock time = 0.000164

**Comment by zhenz** [comment]
@raffenet I just rebuilt the images and updated https://www.chameleoncloud.org/appliances/71/. Ran cpi example successfully. Let me know if everything's good to you as well. If so I'll publish the appliance.

**Comment by raffenet** [comment]
Looks good to me. 👍

**Pull request merged**